hub: add retention policy

jperezdealgaba opened the pull request:
I implemented the retention policy for task results files as specified on: https://issues.redhat.com/browse/OSH-19?filter=-1

lzaoral commented:
Thank you for your contribution, @jperezdealgaba!  I've marked the PR as draft because there is still some necessary work to do to make the CI green and thus the PR ready for review.  

Also, please change the PR name and commit messages and lint every commit to adhere to our contribution guidelines: https://github.com/openscanhub/openscanhub/blob/main/CONTRIBUTING.md Thank you!

jperezdealgaba commented:
I just added the precommit dependencies and did a new commit where the linting errors should be fixed

lzaoral commented:
@jperezdealgaba The Ci is still failing with the following traceback:

```python
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.ProgrammingError: relation "scan_retentionpolicysetting" does not exist
LINE 1: ...egory", "scan_retentionpolicysetting"."days" FROM "scan_rete...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/lib/python3.6/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "scan_retentionpolicysetting" does not exist
LINE 1: ...egory", "scan_retentionpolicysetting"."days" FROM "scan_rete...
                                                             ^


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "osh/hub/manage.py", line 13, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/commands/test.py", line 23, in run_from_argv
    super().run_from_argv(argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3.6/site-packages/django/core/management/commands/test.py", line 55, in handle
    failures = test_runner.run_tests(test_labels)
  File "/usr/lib/python3.6/site-packages/django/test/runner.py", line 725, in run_tests
    old_config = self.setup_databases(aliases=databases)
  File "/usr/lib/python3.6/site-packages/django/test/runner.py", line 645, in setup_databases
    debug_sql=self.debug_sql, parallel=self.parallel, **kwargs
  File "/usr/lib/python3.6/site-packages/django/test/utils.py", line 183, in setup_databases
    serialize=connection.settings_dict['TEST'].get('SERIALIZE', True),
  File "/usr/lib/python3.6/site-packages/django/db/backends/base/creation.py", line 90, in create_test_db
    self.connection._test_serialized_contents = self.serialize_db_to_string()
  File "/usr/lib/python3.6/site-packages/django/db/backends/base/creation.py", line 136, in serialize_db_to_string
    serializers.serialize("json", get_objects(), indent=None, stream=out)
  File "/usr/lib/python3.6/site-packages/django/core/serializers/__init__.py", line 129, in serialize
    s.serialize(queryset, **options)
  File "/usr/lib/python3.6/site-packages/django/core/serializers/base.py", line 90, in serialize
    for count, obj in enumerate(queryset, start=1):
  File "/usr/lib/python3.6/site-packages/django/db/backends/base/creation.py", line 133, in get_objects
    yield from queryset.iterator()
  File "/usr/lib/python3.6/site-packages/django/db/models/query.py", line 353, in _iterator
    yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
  File "/usr/lib/python3.6/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/usr/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1178, in execute_sql
    cursor.close()
psycopg2.OperationalError: cursor "_django_curs_140450644335488_sync_30" does not exist
```

Since you have added a now model, you have to also create and commit a corresponding Django migration.

lzaoral commented:
Well, that was fast 😄.

jperezdealgaba commented:
@lzaoral I just uploaded the migrations file. I thought they were generated automatically

jperezdealgaba commented:
Finally. We can start the review!

lzaoral reviewed (changes requested):
Thank you for fixing the code to make the CI green, @jperezdealgaba!

* Could you please create a separate script and a systemd unit with a timer for removal of the old logs?  Right now `osh-stats` is used to gather scan statistics and this feature is not related to it in any way (apart from using a timer to be executed periodically) so mixing them would definitely confuse our users.

* The commit messages and the PR name are still not following our contribution guidelines.  Currently, the commit messages are without an prefix and OSH-19 is not on the list of our prefixes.  Generally the prefix says *where* are the changes done, not *what* are they changing. 

* Some commits are fixing previous commits in the PR.  This is generally frowned upon because it needlessly complicates the Git history and in this case also breaks the rule that every separate commit should not make the CI red in any way.  Moreover, this complicates reviewing of such PRs because the reviewer might find some issue introduced in some commit which will be fixed in some future commit and it also complicates reverting of such commits (you will have to revert a sequence of intertwined commits instead of a single one).  If you're not sure how to restructure the Git history, I recommend reading this short chapter from ProGit: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

Thank you!

jperezdealgaba commented:
Hey @lzaoral ! I think I fixed all the things you commented on last comment. I squashed the commits, changed commits and PR names and created a new service to enforce the policy retention

lzaoral commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,23 @@ def csmock_args(self):
             return ''
+
+
+class RetentionPolicySettingsManager(models.Manager):
+    def get_parameter(self, category):
+        try:
+            days = self.get(category=category)
+        except ObjectDoesNotExist:
+            logger.error("category %s does not exist", category)
+            raise ObjectDoesNotExist("category %s does not exist" % category)
+        return days
+
+
+class RetentionPolicySetting(models.Model):
+    category = models.CharField(max_length=128, blank=False, null=False)
+    days = models.PositiveIntegerField(
+        default=14,
+        help_text="Number of days before the text is deleted "
+    )
+
+    objects = RetentionPolicySettingsManager()
```
Right now, there are only two elements in the corresponding fixture.  If we definitely plan to add other settings in the future, having a separate model is a good idea.  If we don't, I suggest to use the `AppSettings` model instead which already contains, e.g. the number of days after which a not waived scan is overdue.

jperezdealgaba replied:
For this, I had a discussion with @kdudka and we decided to create a new model for this.That's why we didn't use the AppSetttings model.

lzaoral replied:
Cool!  Thank you for the explanation!

kdudka replied:
Yes, we decided to use a separate table because we are going to append additional rows in the future.  Failed scans and user-initiated scans are the low-hanging fruit.  But for scans initiated using a service principal, we need to distinguish scans in CI vs. scans of production builds.  For production builds we need to distinguish the ones that were in the end shipped to customers vs. the builds that were overridden by a subsequent build, etc.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+    for scan in scansToDelete:
+        try:
+            pathToDelete = models.Task.get_task_dir(scan.id)
+            if os.path.isdir(pathToDelete):
+                shutil.rmtree(pathToDelete)
```
Please make the directory removal unconditional because right now it introduces a TOCTOU bug (https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use).  If the target path does not exist, it will be caught by the `except` statement anyway.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+            if os.path.isdir(pathToDelete):
+                shutil.rmtree(pathToDelete)
+        except OSError as e:
+            logging.error(f"Error deleting scans: {e}")
```
While inconsistent, the `logger` model from Python's standard library uses a `printf`-like formatting:
```suggestion
            logging.error("Error deleting scans: %s", e)
```

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+    django.setup()
+
+    # Retention policy Scans
+    print("Starting Retention Policy Enforcement")
```
This should be logged and not printed to the standard output.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+        retentionPersonalObject = models.RetentionPolicySetting.objects.get(category="PersonalScan")
+        retentionPersonalScans = retentionPersonalObject.days
+    else:
+        # We set up the default time
+        retentionPersonalScans = 14
```
This variable was already set to the same value on line 51.  Thus, this assignment is redundant.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+
+
+# In order to identify personal scans, we are using the owner field. If the owner doesn't contain any / or @, we assume
+# that is a personal scan
```
Typo and a missing full stop:

```suggestion
# that it is a personal scan.
```

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+
+# In order to identify personal scans, we are using the owner field. If the owner doesn't contain any / or @, we assume
+# that is a personal scan
+def identifyPersonalScan(owner):
+    # We could use regexps here if we want to make this more complex but for this specific case
+    # built-in string methods are enough and we are more perfomance-oriented.
+    if "/" in owner or "@" in owner:
+        return False
+    else:
+        return True
```
This whole expression can be simplified as
```python
return "/" not in owner and "@" not in owner
```
which is rather short for a separate function.  Thus, it can be easily inlined to the calling function.

jperezdealgaba replied:
In the new version, I used your return statement but I would like to keep the function as it helps to identify which rules are we checking. In the future when we have more rules, It would be good to be able to identify them quickly and add more cases

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+
+    # ---------- Section to take parameters from DB
+
+    retentionPersonalScans = 14
+    retentionFailedStatus = 14
+
+    if models.RetentionPolicySetting.objects.filter(category="PersonalScan").exists():
+        retentionPersonalObject = models.RetentionPolicySetting.objects.get(category="PersonalScan")
+        retentionPersonalScans = retentionPersonalObject.days
+    else:
+        # We set up the default time
+        retentionPersonalScans = 14
+        logging.warning("Error getting PersonalScan parameter. Setting default parameter")
+
+    if models.RetentionPolicySetting.objects.filter(category="FailedStatus").exists():
+        retentionFailedObject = models.RetentionPolicySetting.objects.get(category="FailedStatus")
+        retentionFailedStatus = retentionFailedObject.days
+    else:
+        # We set up the default time
+        retentionFailedStatus = 14
+        logging.warning("Error getting FailedStatus parameter. Setting default parameter")
+
+    # ---------- End Section to take parameters from DB
```
The code blocks that retrieve the values from the database are nearly similar.  Therefore, they are perfect candidates to be factored-out to a single function that will get and process a single object from the database.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+        retentionFailedStatus = retentionFailedObject.days
+    else:
+        # We set up the default time
+        retentionFailedStatus = 14
```
This variable was already set to the same value on line 52. Thus, this assignment is redundant.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+                eligiblePersonalScans.append(t)
+        elif calculatedDateStatusFailed <= today:
+            # Failed status eligible to be deleted. Status 5 means to be deleted:
+            if t.state == 5:
```
`5` is a magic number.  Please use the corresponding failed task state tuple from `kobo`: https://github.com/release-engineering/kobo/blob/986f2a305f9c675dbb54e34c91f09e443e328af3/kobo/client/constants.py#L36

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+    # We identify which scans are eligible to be deleted
+    today = date.today()
+    for t in tasks:
+        # At the moment, we only have 2 rule for retention settings:
```
Typo:
```suggestion
        # At the moment, we only have 2 rules for retention settings:
```

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+        # Failed status
+        calculatedDatePersonalScans = (t.dt_created + datetime.timedelta(days=retentionPersonalScans)).date()
+        calculatedDateStatusFailed = (t.dt_created + datetime.timedelta(days=retentionFailedStatus)).date()
+        if calculatedDatePersonalScans <= today:
+            if identifyPersonalScan(t.owner.username):
+                # Personal Scans eligible to be deleted
+                eligiblePersonalScans.append(t)
+        elif calculatedDateStatusFailed <= today:
```
I believe we want to remove non-personal tasks that that are older than `calculatedDatePersonalScans`, right?  This is not the case with the current code, though.  To fix it, I'd use a following code structure:

```python
if <task is personal>:
    <check and mark personal task for removal>
    continue

<check and mark non-personal task for removal>
```

jperezdealgaba replied:
I don't really understand your comment here. Nevertheless, I made the code clearer:

if <task is personal>:
    <check and mark personal task for removal>
if <task is failed>:
<check and mark failed task for removal>

In previous comment, we checked before if it could be deleted because of the time constaints and then we would check the category.

lzaoral replied:
Thank you!  Your recent changes actually fixed the bug I was referring to!  Just for the sake of completeness:

```python
if calculatedDatePersonalScans <= today:
    if identifyPersonalScan(t.owner.username):
        ...
elif calculatedDateStatusFailed <= today:
        ...
```
Due to the `elif` a failed non-personal task that satisfied the personal task limit would not be deleted even when it also satisfies the failed limit.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,99 @@
+    deleteScansResults(eligiblePersonalScans)
+    deleteScansResults(eligibleFailedStatus)
+
+    # --- End section-----
```
I'd be nice to also log the successful completion of this script.

lzaoral reviewed (changes requested):
Thank you, @jperezdealgaba!  This looks much better already!

Specific comments were left below, general comments follow:

* In OSH, `Scan` and `Task` are separate models.  A task may exist without a scan and scan is generally associated with some set of tasks.  Please, update the identifiers (variables, functions, ...) accordingly because this code operates on the `Task` model but refers to its objects as scans which is confusing.

lzaoral commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,26 @@ def csmock_args(self):
+class RetentionPolicySettingsManager(models.Manager):
+    def get_days_from_category(self, category):
+
+        # We establish a broad default days, so we don't delete unwanted results.z
```
Typo:
```suggestion
        # We establish a broad default days, so we don't delete unwanted results.
```

lzaoral commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,26 @@ def csmock_args(self):
+    category = models.CharField(max_length=128, blank=False, null=False)
+    days = models.PositiveIntegerField(
+        default=14,
+        help_text="Number of days before the text is deleted "
```
Sorry, missed this one in the previous round.  What `"text"` is the field description referring to?  I believe you meant logs, task artifacts, etc.  Also, the description string contains a trailing white-space character.

lzaoral commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,26 @@ def csmock_args(self):
+            days = self.get(category=category).days
+        except ObjectDoesNotExist:
+            logger.warning("Category %s could not be found. Setting default parameter", category)
+            days = 60
```
Previous iteration had `14` as the default number of days for an unknown category.  Is the change to `60` expected?

lzaoral replied:
Sure, makes perfect sense.  I just wanted to be sure it is not a typo.  Thank you!

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,82 @@
+                # Personal Scans eligible to be deleted
+                eligiblePersonalTasks.append(t)
+        elif t.is_failed():
+            # Failed status eligible to be deleted. Status 5 means to be deleted:
```
This comment seems redundant now.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,82 @@
+        calculatedDatePersonalTasks = (t.dt_created + datetime.timedelta(days=retentionPersonalTasks)).date()
+        calculatedDateStatusFailed = (t.dt_created + datetime.timedelta(days=retentionFailedStatus)).date()
+
+        if identifyPersonalTask(t.owner.username):
+            if calculatedDatePersonalTasks <= today:
```
Generally two completely tested `if` statements can be combined into a single one using `and`.  Doing so makes control flow more simple and it also cuts-down code indentation:
```suggestion
        if identifyPersonalTask(t.owner.username) and calculatedDatePersonalTasks <= today:
```

The same holds also for the nested `elif` and `if` below.

lzaoral reviewed (changes requested):
I missed this one in the previous round.  The function and variable identifiers are not adhering to PEP8: https://peps.python.org/pep-0008/#function-and-variable-names

jperezdealgaba commented:
@lzaoral Could you please check the changes? I think I fixed all of them

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+os.environ['DJANGO_SETTINGS_MODULE'] = 'osh.hub.settings'
+
+
+def delete_tasks_results(tasksToDelete):
```
The `tasksToDelete` identifier is not adhering to PEP8 variable naming guidelines.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+    from kobo.hub import models
+    for task in tasksToDelete:
+        try:
+            pathToDelete = models.Task.get_task_dir(task.id)
```
The `tasksToDelete` identifier is not adhering to PEP8 variable naming guidelines.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+            pathToDelete = models.Task.get_task_dir(task.id)
+            shutil.rmtree(pathToDelete)
+        except OSError as e:
+            logging.error("Error deleting tasks folder: %s", e)
```
There seems to be another typo I missed earlier:
```suggestion
            logger.error("Error deleting tasks folder: %s", e)
```
While this code works, using a specific `logger` instance will also use a correct file information.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+    eligible_personal_tasks = []
+    eligible_failed_status = []
+
+    # We identify which scans are eligible to be deleted
```
`scan` -> `task results`
```suggestion
    # We identify which task results are eligible to be deleted
```

lzaoral commented on osh/hub/osh-retention.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=OpenScanHub retention policy enforcer
+StartLimitBurst=2
+StartLimitIntervalSec=2h
```
I'll keep this conversation open for now so that we can tweak the values further after we test it on the staging and production instances.

lzaoral replied:
Conversation moved to https://github.com/openscanhub/openscanhub/pull/76#issuecomment-1723337676.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+
+    # ---------- End Section to take parameters from DB
+
+    eligible_personal_tasks = []
+    eligible_failed_status = []
```
As suggested by @kdudka in private, we can just use a single array to store the to-be-deleted tasks to simplify the code further.

lzaoral reviewed (changes requested):
Thank you, @jperezdealgaba!  Please, see comments below.

jperezdealgaba commented:
I uploaded the new changes.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,79 @@
+            eligible_tasks_to_b_deleted.append(t)
+
+    delete_tasks_results(eligible_tasks_to_b_deleted)
+    delete_tasks_results(eligible_tasks_to_b_deleted)
```
This line can be safely removed thanks to https://github.com/openscanhub/openscanhub/pull/76#discussion_r1274901833.

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,79 @@
+
+    # ---------- End Section to take parameters from DB
+
+    eligible_tasks_to_b_deleted = []
```
Typo: `eligible_tasks_to_b_deleted` -> `eligible_tasks_to_be_deleted`

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,79 @@
+    logger.info("Starting Retention Policy Enforcement")
+
+    from osh.hub.scan import models
+    tasks = models.Task.objects.all()
```
On a second thought, we definitely want to apply this policy only to finished tasks:
```python
from kobo.client.constants import FINISHED_STATES
tasks = models.Task.objects.filter(state__in=FINISHED_STATES)
```

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,79 @@
+        # At the moment, we only have 2 rules for retention settings:
+        # Personal tasks
+        # Failed status
+        calculated_date_personal_tasks = (t.dt_created + datetime.timedelta(days=retention_personal_tasks)).date()
+        calculated_date_status_failed = (t.dt_created + datetime.timedelta(days=retention_failed_status)).date()
```
@kdudka Do we want to compute the date from the moment of creation or from the moment the task was finished?

jperezdealgaba replied:
I used the creation date to make sure that we can't also delete tasks that are stuck, don't find a worker,.... But I don't have any problem changing it to the finished date. Please, lmk so I can add it in the next commit.

lzaoral replied:
The concern you have is unrelated to the comment above.  All tasks (even unassigned or stuck) have a creation date set immediately when they are saved to the database:

https://github.com/release-engineering/kobo/blob/986f2a305f9c675dbb54e34c91f09e443e328af3/kobo/hub/models.py#L560C6-L560C6

To exclude such tasks, we need to implement https://github.com/openscanhub/openscanhub/pull/76#discussion_r1274951302 first.

jperezdealgaba replied:
Cool. Just pushed those changes

jperezdealgaba commented:
I fixed the rest with the exception of https://github.com/openscanhub/openscanhub/pull/76#discussion_r1275065094 until something is decided

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+            path_to_delete = models.Task.get_task_dir(task.id)
+            shutil.rmtree(path_to_delete)
+        except OSError as e:
+            logger.error("Error deleting tasks folder: %s", e)
```
Do we record which tasks (in fact task-specific directories) have already been deleted?

If not we will hit `FileNotFoundError` exceptions for the already deleted tasks on each subsequent run.  It is also wasteful to iterate over 200000 directories each day, just to discover that they do not exist any more.  I think we should introduce a task-specific flag to mark the tasks with deleted data and exclude them in the initial db query already.

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,80 @@
+
+        if identify_personal_task(t.owner.username) and calculated_date_personal_tasks <= today:
+            # Personal Scans eligible to be deleted
+            eligible_tasks_to_be_deleted.append(t)
```
We need to log log which tasks have been deleted on which criteria.  If we discover that some data is unexpectedly missing one day, there should be a way to audit what exactly happened and why.

lzaoral replied:
I believe that this can be now implemented with a `ForeignKey` in `TaskResultsRemoval` pointing to `RetentionPolicySetting`.

jperezdealgaba replied:
I added it. Thanks for the comment

kdudka reviewed (changes requested):
Nice work.  Thank you both!  Two comments inline...

jperezdealgaba commented:
@lzaoral Hey Lukas! I have implemented all the changes and I have modified the task model in order to implement the flag that determines the retention policy has been applied. This was changed in kobo so I guess until that PR is not accepted, this tests will fail. (I wasn't able to determine if the process is different for modifying kobo). The PR for kobo is: https://github.com/release-engineering/kobo/pull/211

lzaoral commented:
Thank you!  This PR is also blocked on https://github.com/release-engineering/kobo/issues/49 which must be fixed first.  Otherwise, the client would be able to refuse to download results that we're removed by `osh-retention`.

lzaoral commented:
Marking as draft until the PRs in `kobo` are resolved.

jperezdealgaba commented:
Sure. It makes sense

jperezdealgaba commented:
I have just done a new commit where the property `retention_applied` is not used and we create a new model where the already "retentioned" tasks are stored. In that way, we don't modify kobo

jperezdealgaba commented:
Update: I have removed the draft status

lzaoral reviewed (changes requested):
Thanks for the update, @jperezdealgaba!  Could you please remove the existing migrations and generate a fresh one?  There is no need to create six migrations for a completely new Django model.

See inline suggestions below.

rhyw commented on osh.spec:
```diff
@@ -241,12 +243,15 @@ fi
 pg_isready -h localhost && %{python3_sitelib}/osh/hub/manage.py migrate
 
 %systemd_post osh-stats.service osh-stats.timer
+%systemd_post osh-retention.service osh-retention.timer
 
 %preun hub
 %systemd_preun osh-stats.service osh-stats.timer
+%systemd_preun osh-retention.service osh-retention.timer
 
 %postun hub
 %systemd_postun_with_restart osh-stats.service osh-stats.timer
+%systemd_postun_with_restart osh-retention.service osh-retention.timer
```
I don't think it makes much sense to maintain a sorted order for a list in `setup.py` or here in a SPEC file.

kdudka replied:
@rhyw If you look for a string in a sorted list it is much faster (logarithmic vs. linear complexity).  Also if you want to see differences between two lists, check for missing items, etc. it is always easier if the lists are already sorted.

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,31 @@ def csmock_args(self):
+    objects = RetentionPolicySettingsManager()
+
+
+class RetentionScan(models.Model):
```
What about something more specific such as `RetentionAppliedTask` ?

kdudka replied:
I am not sure how much OpenScanHub is following Django's best practices but I think the name should start with `Task` because it is bound to the `Task` table, e.g. `TaskRemoval`, or `TaskPurge` (so that a plural of it makes sense).  There seem to be preceding cases with `Package -> PackageAttribute`, `Scan -> ScanBinding`, `Result -> ResultGroup`, etc.

I think we should also record a timestamp when the task data was removed so that we have the answer when users ask for it (and so that we can verify that the retention policy implementation really follow its configuration).

jperezdealgaba replied:
Hey! I modified the names of this and created the `TaskResultsRemoval` model. Also, I have added a new field with the datetime when the task is added to the table

jperezdealgaba commented on osh.spec:
```diff
@@ -209,9 +209,11 @@ fi
 %files hub
 %defattr(-,root,apache,-)
 %{_sbindir}/osh-stats
+%{_sbindir}/osh-retention
```
Fixed

jperezdealgaba commented on osh.spec:
```diff
@@ -209,9 +209,11 @@ fi
 %{_sysconfdir}/osh/hub
 %{python3_sitelib}/osh/hub
 %{_unitdir}/osh-stats.*
+%{_unitdir}/osh-retention.*
```
Fixed

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -22,23 +22,22 @@
-                         #  (something went wrong during build process or
-                         #  analyser had some problems)
-    "BASE_SCANNING",     # child scan is in scanning process right now
-    "CANCELED",          # there is newer build submitted, this one is obsolete
-    "DISPUTED",          # scan was waived but one of waivers was obsoleted
-    "INIT",              # first, default state
-    "BUG_CONFIRMED",     # run contains at least one group marked as bug
+    #  owner's attention
+    "WAIVED",  # user appropriately waived each defect
+    "PASSED",  # scan didn't discover new defects; everything is fine
+    "FINISHED",  # scan finished -- USER/ERRATA_BASE scans only
+    "FAILED",  # scan has failed, need an attention by OpenScanHub admins
+    #  (something went wrong during build process or
+    #  analyser had some problems)
+    "BASE_SCANNING",  # child scan is in scanning process right now
+    "CANCELED",  # there is newer build submitted, this one is obsolete
+    "DISPUTED",  # scan was waived but one of waivers was obsoleted
+    "INIT",  # first, default state
+    "BUG_CONFIRMED",  # run contains at least one group marked as bug
 )
```
Fixed

lzaoral replied:
I can still see the formatting changes in the diff...

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,105 @@
+        # If the tasks results are applied, we update the database field, so we don't try to delete them again.
+        try:
+            retention_scan = osh.hub.scan.models.RetentionScan(task=task)
+            retention_scan.save()
```
Fixed

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,105 @@
+        tasks = models.Task.objects.filter(state__in=FINISHED_STATES).exclude(id__in=retention_scans)
+    except DatabaseError as e:
+        logger.error(f"Error getting retention_tasks: {str(e)}")
+        tasks = []
```
Fixed

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,105 @@
+
+    try:
+        # Retrieval of all tasks ids that has already been applied the retention policy
+        retention_scans = models.RetentionScan.objects.values('task').values_list('task', flat=True)
```
Fixed

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -109,6 +108,7 @@ class Permissions(models.Model):
     """
     Custom permissions
     """
+
```
Fixed

lzaoral replied:
I can still see the formatting changes in the diff.

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -145,7 +145,7 @@ class MockConfig(models.Model):
 
     class Meta:
-        ordering = ("name", )
+        ordering = ("name",)
```
Fixed

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -337,17 +337,17 @@ def display_graph(self, scan, response, indent_level=0):
-                               .values_list('release', flat=True)
+        blocked_releases = self.get_partially_blocked_releases() \
+            .values_list('release', flat=True)
         scans = Scan.objects.filter(package=self)
 
         if not blocked_releases and not scans:
             return mark_safe('There are no scans submitted related to this \
 package')
 
-        releases = scans.filter(tag__release__isnull=False)\
-                        .values_list('tag__release', flat=True)\
-                        .union(blocked_releases)
+        releases = scans.filter(tag__release__isnull=False) \
+            .values_list('tag__release', flat=True) \
+            .union(blocked_releases)
         response = ""
```
Fixed

lzaoral replied:
I can still see the formatting changes in the diff.

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -861,11 +861,11 @@ class ScanBinding(models.Model):
     result = models.OneToOneField("waiving.Result", on_delete=models.CASCADE,
-                                  blank=True, null=True,)
+                                  blank=True, null=True, )
 
```
Fixed

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -1013,7 +1013,7 @@ def setting_check_user_can_submit(cls):
          "Y" => hub has to check user perm 'scan.errata_xmlrpc_scan'
          "N" => hub does not have to check
         """
-        return cls.objects.get(key="CHECK_USER_CAN_SUBMIT_SCAN").\
+        return cls.objects.get(key="CHECK_USER_CAN_SUBMIT_SCAN"). \
             value.upper() == "Y"
```
Fixed

lzaoral replied:
I can still see the formatting changes in the diff.

jperezdealgaba commented on setup.py:
```diff
@@ -36,6 +38,7 @@
     "/usr/sbin": [
         "osh/hub/scripts/osh-stats",
+        "osh/hub/scripts/osh-retention",
         "osh/worker/osh-worker",
```
Fixed

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,31 @@ def csmock_args(self):
+
+class RetentionScan(models.Model):
+    task = models.ForeignKey(Task, on_delete=models.PROTECT,
+                             help_text="Scan that has been applied retention policy")
```
Fixed

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,105 @@
+def delete_tasks_results(tasks_to_delete):
+    from kobo.hub import models
+
+    import osh.hub.scan.models
```
Hey, I have moved to this to the top and make the models explicit, but I needed to add the `# noqa` comment as the pre-commit will complaint about some imports not at the top of the documen. The django config needs to be defined before the models are imported. Lmk if there is a better way to do that

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,105 @@
+            retention_scan.save()
+        except DatabaseError as e:
+            logger.error("Error updating database value for task %d: ", task.id)
+            logger.error("Exception raised in the database %s: ", e)
```
Fixed

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,105 @@
+
+    from kobo.client.constants import FINISHED_STATES
+
+    from osh.hub.scan import models
```
Fixed

jperezdealgaba replied:
Ahh sorry. I misinterpeted it. I just fixed it

lzaoral commented on osh/hub/scan/models.py:
```diff
@@ -746,7 +746,7 @@ def get_first_scan_binding(self):
             scan__package=self.package,
             scan__tag__release=self.tag.release,
             task__state=TASK_STATES['CLOSED'],
-            scan__scan_type__in=SCAN_TYPES_TARGET).\
+            scan__scan_type__in=SCAN_TYPES_TARGET). \
             order_by('result__date_submitted')
```
I can still see the formatting changes in the diff.

jperezdealgaba replied:
This should be fixed

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,89 @@
+from osh.hub.scan.models import RetentionPolicySetting  # noqa
+from osh.hub.scan.models import TaskResultsRemoval  # noqa
+
+logger = logging.getLogger("osh.hub.scripts.osh-retention")
```
Is there a reason why we don't use `__name__` here as we do in the rest of the project?
```suggestion
logger = logging.getLogger(__name__)
```

jperezdealgaba replied:
I used that instead of `__name__` because at the time of logging, logging was not done correctly and there was no output. At the time of using that, everything was fixed. I also saw some other modules use a separate logging, such as `logger = logging.getLogger('osh_api_client')`

kdudka replied:
@jperezdealgaba `osh/hub/scripts/osh-xmlrpc-client.py` is a debugging script that we do not install, so it might have some technical debt.

jperezdealgaba replied:
I have thinking and trying about this. The rest of the files that uses `__name__` doesn't have a main function as the osh-retention does. even the `osh-stats` uses another module, that makes that using `__name__` will resolve in the correct file. `osh-retention` uses one file. I can separate the modules and create two, but I don't think here it is necessary.

kdudka commented on osh/hub/scan/fixtures/initial_data.json:
```diff
@@ -2724,5 +2724,21 @@
+  "model": "scan.retentionpolicysetting",
+  "fields": {
+    "category": "PersonalScan",
+    "days": 7
```
I would set the retention to 30 days by default.

jperezdealgaba replied:
Done

kdudka commented on osh/hub/scan/fixtures/initial_data.json:
```diff
@@ -2724,5 +2724,21 @@
+  "model": "scan.retentionpolicysetting",
+  "fields": {
+    "category": "FailedStatus",
+    "days": 1
```
I would set the retention to 30 days by default.

jperezdealgaba replied:
Done

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,89 @@
+    tasks = Task.objects.filter(state__in=FINISHED_STATES).exclude(id__in=processed_tasks)
+
+    # ---------- Section to take parameters from DB
+    retention_personal_tasks = RetentionPolicySetting.objects.get(category="PersonalScan")
```
We should have a try/except block here, so that we log an error if the configuration is missing rather than printing traceback to `stderr`:
```
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com systemd[1]: Starting OpenScanHub retention policy enforcer...
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]: Traceback (most recent call last):
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:   File "/usr/sbin/osh-retention", line 89, in <module>
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:     main()
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:   File "/usr/sbin/osh-retention", line 61, in main
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:     retention_personal_tasks = RetentionPolicySetting.objects.get(category="PersonalScan")
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:   File "/usr/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:     return getattr(self.get_queryset(), name)(*args, **kwargs)
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:   File "/usr/lib/python3.6/site-packages/django/db/models/query.py", line 437, in get
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]:     self.model._meta.object_name
Sep 07 13:13:59 covscan.lab.eng.brq2.redhat.com osh-retention[3894367]: osh.hub.scan.models.DoesNotExist: RetentionPolicySetting matching query does not exist.
Sep 07 13:14:00 covscan.lab.eng.brq2.redhat.com systemd[1]: osh-retention.service: Main process exited, code=exited, status=1/FAILURE
Sep 07 13:14:00 covscan.lab.eng.brq2.redhat.com systemd[1]: osh-retention.service: Failed with result 'exit-code'.
```

jperezdealgaba replied:
Done

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,89 @@
+        # At the moment, we only have 2 rules for retention settings:
+        # Personal tasks
+        # Failed status
+        calculated_date_personal_tasks = (t.dt_finished + datetime.timedelta(days=retention_personal_tasks.days)).date()
```
When I tried to run the script on `covscan.lab.eng.brq2.redhat.com`, it failed with the following traceback before it deleted any results:
```py
# /usr/sbin/osh-retention
Traceback (most recent call last):
  File "/usr/sbin/osh-retention", line 89, in <module>
    main()
  File "/usr/sbin/osh-retention", line 73, in main
    calculated_date_personal_tasks = (t.dt_finished + datetime.timedelta(days=retention_personal_tasks.days)).date()
TypeError: unsupported operand type(s) for +: 'NoneType' and 'datetime.timedelta'
```

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,89 @@
+def delete_tasks_results(tasks_to_delete):
+    for task, retention_policy_setting in tasks_to_delete:
+        try:
+            logger.info('The following task results will be deleted: %s.', task.id)
```
Done

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,89 @@
+            logger.info('The following task results will be deleted: %s.', task.id)
+            path_to_delete = Task.get_task_dir(task.id)
+            shutil.rmtree(path_to_delete)
+            logging.info('The following folder was deleted:  %s', path_to_delete)
```
Done

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,89 @@
+def identify_personal_task(owner):
+    # We could use regexps here if we want to make this more complex but for this specific case
+    # built-in string methods are enough, and we are more perfomance-oriented.
+    return "/" not in owner and "@" not in owner
```
We will need to decide on this

lzaoral commented:
@kdudka Could you please check that the values in `osh-retention.{service,timer}` are suitable?  Namely the repetition timing and the `StartLimit*` values.

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+
+os.environ['DJANGO_SETTINGS_MODULE'] = 'osh.hub.settings'
+django.setup()
+from kobo.hub.models import Task  # noqa
+
+from osh.hub.scan.models import RetentionPolicySetting  # noqa
+from osh.hub.scan.models import TaskResultsRemoval  # noqa
```
@lzaoral See https://github.com/openscanhub/openscanhub/pull/76#discussion_r1307918886.

kdudka commented:
> @kdudka Could you please check that the values in `osh-retention.{service,timer}` are suitable? Namely the repetition timing and the `StartLimit*` values.

@lzaoral I do not think we need `Restart=on-failure` in the first place (neither for `osh-retention` nor for `osh-stats`).

jperezdealgaba commented:
@kdudka I have added the `-` validation on the owner field and added the specific `noqa` comment

hanchuntao commented on containers/scripts/generate_integration_test_coverage.sh:
```diff
@@ -132,6 +132,12 @@ main() {
+    podman exec osh-hub /usr/bin/coverage-3.6 run --parallel-mode '--omit=*site-packages*,*kobo*,' --rcfile=/coveragerc osh/hub/scripts/osh-retention
+
+    # test osh-retention policy enforcer
+    podman exec osh-hub /usr/bin/coverage-3.6 run --parallel-mode '--omit=*site-packages*,*kobo*,' --rcfile=/coveragerc osh/hub/scripts/osh-retention
```
Why does this command run twice?

jperezdealgaba commented:
@hanchuntao The double command was fixes on latest commit

hanchuntao commented:
> @hanchuntao The double command was fixes on latest commit

got it, thanks for your update

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+            logging.debug('The following folder was deleted: %s', path_to_delete)
+        except OSError as e:
+            logger.error("Error deleting tasks folder: %s", e)
+            return
```
`return` -> `continue`

If the data is for whatever reason deleted from the file system already, it is not a reason to entirely give up on the retention policy a skip removing the other 200 000 tasks that should be deleted.

kdudka replied:
Actually, we need to distinguish two cases:

1. the data is already deleted - then we should mark as deleted in the database
2. the data still exists but cannot be deleted - then we should try it next time again

jperezdealgaba replied:
Modified

lzaoral reviewed (changes requested):
Some pending remarks that I have already written few weeks back are above.

Please, rebase and regenerate the migrations to reflect recent changes in this PR and also other changes in `main`, e.g.:
```shell
$ podman exec -it osh-hub osh/hub/manage.py makemigrations -v3 --dry-run
Migrations for 'scan':
  osh/hub/scan/migrations/0017_alter_retentionpolicysetting_days.py
    - Alter field days on retentionpolicysetting
Full migrations file '0017_alter_retentionpolicysetting_days.py':
# Generated by Django 3.2.20 on 2023-11-03 18:05

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('scan', '0016_retentionpolicysetting_taskresultsremoval'),
    ]

    operations = [
        migrations.AlterField(
            model_name='retentionpolicysetting',
            name='days',
            field=models.PositiveIntegerField(default=30, help_text='Number of days before the task is deleted.'),
        ),
    ]
```

Also, if you get a traceback, it will be inturrepted by another traceback because `django.core.exceptions` is a python module and not a class:
```
INFO    2023-11-03 18:14:53,128 /src/osh/hub/scripts/osh-retention                :56 main   Starting Retention Policy Enforcement
...

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/src/osh/hub/scripts/osh-retention", line 107, in <module>
    main()
  File "/src/osh/hub/scripts/osh-retention", line 66, in main
    except django.core.exceptions as e:
TypeError: catching classes that do not inherit from BaseException is not allowed
```

I'll test the script further when these issues are resolved.

jperezdealgaba commented on osh/hub/scan/models.py:
```diff
@@ -1283,3 +1283,31 @@ def csmock_args(self):
+
+
+class RetentionPolicySettingsManager(models.Manager):
+    def get_days_from_category(self, category):
```
This function has been deleted

lzaoral replied:
Hmm, perhaps you forgot to push the changes?  I can still see it in the PR diff.

jperezdealgaba replied:
This has been deleted in the new commit

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+from kobo.client.constants import FINISHED_STATES
+
+os.environ['DJANGO_SETTINGS_MODULE'] = 'osh.hub.settings'
+django.setup()
```
Added

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+
+    delete_tasks_results(eligible_tasks_to_be_deleted)
+    logger.info("Finishing Retention Policy Enforcement")
+    # --- End section-----
```
Deleted

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+        # Failed status
+
+        # If dt_finished is not present, the task is being executing at the moment
+        if not t.dt_finished:
```
Modified

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+        retention_personal_tasks = RetentionPolicySetting.objects.get(category="PersonalScan")
+    except django.core.exceptions as e:
+        logger.error("Error obtaining PersonalScan value %s.", e)
+        exit()
```
Modified

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,107 @@
+
+os.environ['DJANGO_SETTINGS_MODULE'] = 'osh.hub.settings'
+django.setup()
+from kobo.hub.models import Task  # noqa
```
Added

jperezdealgaba commented on osh/hub/scan/fixtures/initial_data.json:
```diff
@@ -2724,5 +2724,21 @@
+  "fields": {
+    "category": "FailedStatus",
+    "days": 30
+  }
```
I reformatted it

jperezdealgaba commented on containers/scripts/generate_integration_test_coverage.sh:
```diff
@@ -132,6 +132,9 @@ main() {
     podman exec osh-hub /usr/bin/coverage-3 run --parallel-mode '--omit=*site-packages*,*kobo*,' --rcfile=/coveragerc osh/hub/scripts/osh-stats
 
+    # test osh-retention policy enforcer
+    podman exec osh-hub /usr/bin/coverage-3.6 run --parallel-mode '--omit=*site-packages*,*kobo*,' --rcfile=/coveragerc osh/hub/scripts/osh-retention
```
Modified

lzaoral commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,110 @@
+    # ---------- Section to take parameters from DB
+    try:
+        retention_personal_tasks = RetentionPolicySetting.objects.get(category="PersonalScan")
+    except django.core.exceptions as e:
+        logger.error("Error obtaining PersonalScan value %s.", e)
+        sys.exit(1)
+
+    try:
+        retention_failed_status = RetentionPolicySetting.objects.get(category="FailedStatus")
+    except django.core.exceptions as e:
+        logger.error("Error obtaining FailedStatus value: %s.", e)
```
The wrong exception usage mentioned in my last review comment is still not resolved.

jperezdealgaba replied:
I changed the exceptions on the new commit

lzaoral replied:
Well, you did change the code but it is still broken and it still crashes on my machine... 

```python
...
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/src/osh/hub/scripts/osh-retention", line 110, in <module>
    main()
  File "/src/osh/hub/scripts/osh-retention", line 69, in main
    except django.db.DatababaseError as e:
AttributeError: module 'django.db' has no attribute 'DatababaseError'
```

@jperezdealgaba I don't want to be rude but do you **actually test** any of your changes locally?  You have to understand that OpenScanHub maintenance is not in my job description and I'd love to spend the little time I have for it by doing something actually meaningful and not by repeated testing of your broken code which should have been tested already by yourself before you pushed it to GitHub.

jperezdealgaba replied:
I have been triggering the exception in different ways and added new exception condtion cases to have a better exception management.

lzaoral commented:
Any updates, @jperezdealgaba?  There has been no activity for over a month.

jperezdealgaba commented:
@lzaoral I am not able to answer to your comment directly. I did the changes and the last committed versions include changes I discussed with @kdudka such as not deleting the task if the DB parameter is not defined.
The exceptions were raised and tested

lzaoral reviewed (changes requested):
@jperezdealgaba  Please, rebase the PR.  The current branch contents are too outdated and are no longer compatible with a database schema built on the `main` branch leading to the following Traceback during the application of migrations (fortunately, it is unrelated to your changes):
```python
Traceback (most recent call last):
  File "osh/hub/manage.py", line 13, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 89, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 95, in handle
    executor.loader.check_consistent_history(connection)
  File "/usr/lib/python3.6/site-packages/django/db/migrations/loader.py", line 310, in check_consistent_history
    connection.alias,
django.db.migrations.exceptions.InconsistentMigrationHistory: Migration scan.0007_delete_packagecapability is applied before its dependency errata.0003_auto_20230403_1830 on database 'default'.
```

hanchuntao commented on osh/hub/osh-retention.timer:
```diff
@@ -0,0 +1,8 @@
+[Unit]
+Description=Run retention policy enforcemente every day
```
`enforcement`

jperezdealgaba replied:
Fixed!

hanchuntao commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,125 @@
+
+
+if __name__ == '__main__':
+    main()
```
In this file, some lines are too long, I think it is better less than 100

jperezdealgaba replied:
Fixed!

hanchuntao commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,125 @@
+        elif failed_status_task_parameter_exists and t.is_failed() and calculated_date_status_failed <= today:
+            # Failed Scans eligible to be deleted
+            eligible_tasks_to_be_deleted.append([t, retention_failed_status])
+            logger.debug('The following "Failed Scan" task results will be deleted: %s', t.id)
```
I think L102 ~ L118 can be conbined, such as 
```
        if retention_personal_task_parameter_exists:
            calculated_date_personal_tasks = (t.dt_finished + datetime.timedelta(
                days=retention_personal_tasks.days)).date()

            if (identify_personal_task(
                    t.owner.username) and calculated_date_personal_tasks <= today):
                eligible_tasks_to_be_deleted.append([t, retention_personal_tasks])
                logger.debug('Deleting "Personal Scan" task results: %s', t.id)

        if failed_status_task_parameter_exists and t.is_failed():
            calculated_date_status_failed = (t.dt_finished + datetime.timedelta(
                days=retention_failed_status.days)).date()

            if calculated_date_status_failed <= today:
                eligible_tasks_to_be_deleted.append([t, retention_failed_status])
                logger.debug('Deleting "Failed Scan" task results: %s', t.id)
```

jperezdealgaba replied:
Hey @hanchuntao I understand why you mention that. I though about that while coding it, but when we have more complex rules (one that mixes more than one parameter), that approach wouldn't be valid as we would need to mixed a lot of clauses.
Mainly that's why I did a section of calculating parameters and others to select the tasks to be deleted.

What do you think? I am open to change it if we consider that approach is better

hanchuntao replied:
Hello @jperezdealgaba I just think the code looks obscure. If you think this is appropriate, please ignore my comment

jperezdealgaba replied:
Sure! Completely understandable! Just looking for the best option here. @kdudka what do you think about this?

kdudka replied:
I find the code really difficult to read, especially the `if ... _exists` conditions everywhere and the repeated blocks of code.  I think it should be refactored.

kdudka replied:
Addressed in ae6fede98a5660a9f57c7f4da913632eac7c87d3.

hanchuntao commented:
@jperezdealgaba  Could you add some steps about how to test it? thanks!

hanchuntao commented:
I have tested it, and added testing steps in the JIRA comment. LGTM

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+    logger.info("Starting Retention Policy Enforcement")
+
+    # Retrieval of all tasks ids that already have the retention policy applied
+    processed_tasks = TaskResultsRemoval.objects.values_list('task', flat=True)
```
Do we need to fit the list of all previously processed tasks into the memory despite task ID is a foreign key in the table?  The list of processed tasks will grow indefinitely over time.  The time (and memory) taken by this script might depend on the number of tasks to be deleted but not on the number of tasks that have already been deleted in the previous runs.

jperezdealgaba replied:
It is not necessary. I have modified the code to do only one query and we filter through the tasks that are not linked with the `TaskResultsRemoval` table. 
This also improves performance of the script and we should save memory

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+    tasks = Task.objects.filter(state__in=FINISHED_STATES).exclude(id__in=processed_tasks)
+
+    # ---------- Section to take parameters from DB
+    retention_personal_task_parameter_exists = False
```
Why do we need the extra variables with `_exists` suffix?  If a value is missing, we can simply represent it by `None`.

jperezdealgaba replied:
Agree

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+                       "These tasks will not be deleted.")
+    except DatabaseError as e:
+        logger.error("Error obtaining FailedStatus value: %s", e)
+        sys.exit(1)
```
I dislike the extensively duplicated code.  The value retrieval code should be put to a generic function, which will be called multiple times.  This will make it much easier in case we need to introduce other criteria later on.

jperezdealgaba replied:
Refactored

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+
+# The default value matches with the one defined on /hub/scan/models.py
+DEFAULT_RETENTION_DAYS = 30
+DEFAULT_CATEGORY = "Default Category"
```
They stopeed being used. I have deleted them

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+    for task, retention_policy_setting in tasks_to_delete:
+        try:
+            path_to_delete = Task.get_task_dir(task.id)
+            # if folder exists, we delete it. if not, we mark it as deleted as
```
Modified the references

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+                shutil.rmtree(path_to_delete)
+                logger.debug('The following folder was deleted: %s', path_to_delete)
+        except OSError as e:
+            logger.error("Error deleting tasks folder: %s", e)
```
Modified

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+
+
+def main():
+    # Retention policy Task
```
Dropeed!

jperezdealgaba commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,130 @@
+        sys.exit(1)
+    # ---------- End Section to take parameters from DB
+
+    eligible_tasks_to_be_deleted = []
```
Modified

jperezdealgaba commented:
@kdudka I agree with your comment of code's readibility. After so many reviews it was pretty difficult to understand it and follow it.

I have refactored the whole script to make it more readable and in my opinion, far more scalable and understandable for future requirements.

At the time and running locally, we are deleting over 100.000 tasks in over 3 minutes:

`osh-hub osh/hub/scripts/osh-retention  1.47s user 4.19s system 2% cpu 3:14.53 total`

rhyw commented on osh/hub/scripts/osh-retention:
```diff
@@ -26,96 +25,86 @@ from osh.hub.scan.models import TaskResultsRemoval  # noqa: E402
 
+def is_personal(task):
+    # If the username does not contain /, @ or -, assume that the task is personal.
+    return not re.search('/@-', task.owner.username)
```
do you intend to use `[/@-]` here?

kdudka replied:
@rhyw Good catch!  This would not work as expected.

lzaoral replied:
Thank you, @rhyw!  Addressed in 555753918e1de3fe21e39893db930de5dd5001e2.

rhyw commented on osh/hub/scripts/osh-retention:
```diff
@@ -26,96 +25,86 @@ from osh.hub.scan.models import TaskResultsRemoval  # noqa: E402
+def get_setting(name):
     try:
-        obtained_parameter = RetentionPolicySetting.objects.get(category=parameter)
+        return RetentionPolicySetting.objects.get(name=name)
```
if we want to query based on `name`, we better make the field unique and non-nullable. otherwise there might be `MultipleObjectReturned` or other weird behavior here.

hanchuntao commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,111 @@
+
+def has_eligible_time(task, days_delta):
+    if task.dt_finished is None:
+        logger.warn('Task %d is in state %s and the finish time is not set!',
```
The 'warn' method is deprecated, use 'warning' instead 
https://docs.python.org/3.8/library/logging.html#logging.Logger.warning

lzaoral replied:
Thank you, @hanchuntao!  Addressed in 96ae4d976d70d51dd62e141f0c3ed1ef40d60799.

hanchuntao commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,111 @@
+                       "The affected tasks results will not be deleted.")
+        return None
+    except DatabaseError as e:
+        logger.error(f"Error obtaining {name} value %s.", e)
```
logger.error(f"Error obtaining {name} value: %s.", e)

kdudka replied:
Why do we actually combine f-string with `%s` conversion in a single call to `logger.error()`?

lzaoral replied:
My bad, there is no point in using f-strings with loggers.  Addressed in 9ac6c4e0f8d2da46d773a0479da713350af5e3d0.  I've also changed the message format to be consistent with other `logger.error` calls.

kdudka commented on osh/hub/scripts/osh-retention:
```diff
@@ -0,0 +1,111 @@
+    TaskResultsRemoval.objects.create(task=task, reason=reason)
+
+
+def get_setting(name):
```
I cannot see any problem with `get_setting`.

kdudka reviewed (approved):
Looks good.  Thanks for the update!

rhyw reviewed (approved):
Looks good to me. Thanks for the update.

hanchuntao commented:
LGTM, Thanks for the update!

kdudka commented:
/packit rebuild-failed

kdudka commented:
The `testing-farm:fedora-40-x86_64 — Guest couldn't be provisioned: ...` failure seems unrelated to changes in this pull request.
